New package: root-6.24.06

BenJarg opened the pull request:
- I tested the changes in this PR: YES

- This new package conforms to the [quality requirements](https://github.com/void-linux/void-packages/blob/master/Manual.md#quality-requirements): YES

- I built this PR locally for my native architectures (x86_64-glibc and x86_64-musl)

Duncaen commented on common/shlibs:
```diff
@@ -4094,3 +4094,120 @@ libkaddressbookmergelibprivate.so.5 kdepim-addons-21.08.3_1
 libkmailconfirmbeforedeleting.so.5 kdepim-addons-21.08.3_1
 libkaddressbookprivate.so.5 kaddressbook-21.08.3_1
 libcodec2.so.1.0 codec2-1.0.1_1
+libvalarrayDict.so root-6.24.06_1
```
Those libraries shouldn't be listed here, they are not versioned so there is nothing to gain from that.

BenJarg replied:
I see, removed them from the commit

balejk commented:
I was actually about to package this myself, I just haven't yet found the time to polish the template and test it properly. Please feel free to see my template (branch `root` of my fork of `void-packages`) and to copy off anything you like (in particular, I think the `-Dgnuinstall=ON` `cmake` option might be worth noting). There is also a patch included, which enables ROOT to compile on musl too (well, sort of) and which got merged upstream (for more details please refer to [this PR](https://github.com/root-project/root/pull/9288)).

BenJarg commented:
@balejk  Thanks! I think I'll follow your lead and a few of those things to my template. Good to see it is possible to get it to compile with musl; thanks for the help.

Chocimier commented:
> builtin_afterimage builtin_clang builtin_cling builtin_llvm builtin_lz4 builtin_lzma builtin_nlohmannjson builtin_openui5 builtin_pcre builtin_vdt builtin_xxhash builtin_zstd

As much of system libraries as possible should be used.

BenJarg commented:
@Chocimier Ah ok, I updated this to stop using all builtin libraries except afterimage, nlohmannjson, and openui5, which don't seem to have corresponding packages in xbps-src. Another, cling, is part of root, and it seems to be picky about clang and llvm versions it uses, not wanting to use vanilla releases; the preferred build method seems to be using these bundled versions.

balejk commented:
I believe specifying `cmake` as `hostmakedepends` is unnecessary with the selected build style. I also think that it should suffice to use the `gnuinstall` option, installation and mandir prefixes should be set accordingly automatically.

Making using `ninja` worked fine for me, so I see no reason to override it to use `make` - do you have one?

Have you tried cross-building for any other architecture? These two might be the only officially supported ones, but that does not necessarily mean that it won't work on any other.

And have you disabled the `xrootd` option globally on purpose or did you mean to do it just for musl where it won't build? Also, I see that you have set TBB as a dependency - with that, it might be possible to make `imt` available for musl as well, as the compilation is failing there because of the builtin TBB. I believe `cmake` should detect system TBB automatically if you remove `imt=OFF`. Will you be able to try it?

Last thing that comes to mind is that it might be nice to add a `build_option` for the R bindings (preferably disabled by default though).

BenJarg commented:
@balejk Thanks for the feedback!

You're right; I don't need cmake there. And as for the other prefixes, I think I needed them before I added the `gnuinstall` option, but looks like since I added that it works without them, so I'll remove those

At one point I couldn't get it to build right in xbps for some reason with ninja, but it worked with make. Now though looks like I don't have that same issue anymore, so I'll have it use ninja. 

I briefly tried building for ARM but I ran into a number of issues, and then I read that to get it to compile there, a number of changes need to made. I'm not sure if that is still true, I may need to play around with it more to get it to compile there, but I didn't have it work out of the box and read that it wasn't officially supported so I only listed the officially supported architectures.

I disabled `xrootd` globally because I thought a better solution to ROOT building it would be having it as its own package that ROOT depends on (and thought about throwing that package together), but until there is I'll only disable it for musl (it looks like I need to add the versioned `xrootd` libraries to shlibs for that to work also, so I'll do that). I'm rebuilding on a musl system now with `imt` enabled and the other above changes, so hopefully that will work.

And the R bindings build option would be sensible, so I'll add that as well. 

Once I add the above (and see that the musl compilation works with those changes), I'll force push them.

BenJarg commented:
I added all the changes except the option for the R bindings, as one of the required R packages to build them isn't in xpbs (RInside, as far as I can tell anyway).

balejk commented:
[I believe](http://rcpp.org/) RInside should be part of Rcpp (`R-cran-Rcpp` package in `xbps`).

I believe you don't need to specify `make_cmd=ninja` as that should be the default. Also specifying `binutils` is probably unnecessary as it's a dependency of `gcc`.

Regarding XROOTD, indeed it would be better to package it separately and use that (if you got it working on musl, you could even remove the libc restriction), however you still wouldn't want to set `xrootd=OFF`, because if I'm not mistaken, that disables XROOTD completely (both the builtin one and the system one).

BenJarg commented:
@balejk Thanks, I made those adjustments.

As for the R bindings, I've tried adding both `R` and `R-cran-Rcpp` to `makedepends`, and `Dr=ON` in `configure_args`, but still get the error in CMake:
`-- Looking for R
-- Could NOT find R (missing: RInside) 
-- R installation not found. Set variable R_DIR to point to your R installation
-- For the time being switching OFF 'r' option`
so it seems like it's finding the other R components, but not `RInside`. Perhaps I need to point it to it manually?

balejk commented:
Never mind, it seems it's really not part of `xbps`. On the other hand, it should be fairly easy to package it if you were interested as `xbps` is able to fetch packages from CRAN automatically. Or it might be worth asking @wagnerflo or @luhann whether they would do it as they seem to maintain most of the R related packages in `xbps`.

BenJarg commented:
@balejk You're right, CRAN packages are easy to package, so I added `R-cran-RInside` as a new package to this PR (which I think is ok to do in one PR if one package is a dependency of the other, but I'm new to this so let me know if it's not).

Both packages build, but at the end of building ROOT it complains for me that it can't resolve `libRInside.so`(which is from `R-cran-RInside`), which I think is normal because `R-cran-RInside` isn't in xbps yet. I can temporarily fix it by adding `libRInside.so` to shlibs. Again, I'm new to xbps, so someone tell me if there's something else I need to do about that.

balejk commented:
Great! Yes, I believe putting all the related packages in one PR is fine.

I just found out that there are some predefined build options in `xbps` (see [here](https://github.com/void-linux/void-packages/blob/master/common/options.description)), so it would probably make sense to rename the `pyroot` option to `python`.

Regarding the `libRInside.so` problem, does it help if you specify the relevant `R` packages in `depends` (without having them in `shlibs`)?

BenJarg commented:
@balejk Good point, changed pyroot -> python.

And no, I've tried that and it doesn't seem like that changes anything. Still get:
```
SONAME: libRInside.so <-> UNKNOWN PKG PLEASE FIX!
```
I can fix it building in xbps-src by adding it to shlibs, but it still won't install with `xbps-install`, complaining about that library being missing or broken.

BenJarg commented:
This issue may be because libraries for CRAN packages usually are just supposed to be used by R and not used directly by other things, so `libRInside.so` has no soname, i.e. `readelf -d libRInside.so | grep SONAME` doesn't give anything. `libR.so` seems to have this same issue (of package `R`), and I believe this is why `libR.so` is specified in `common/shlibs` even though it isn't versioned. I can resolve these issues by building against the static version of this library instead, but that would probably be an unsatisfactory solution.

BenJarg commented:
Ok, think everything is sorted out now. `libRInside.so` needs to be in shlibs like `libR.so` to be seen right by xbps-src. And everything is found correctly by xbps-install by using `shlib_provides` in the `R-cran-RInside` template.

Chocimier commented:
I like it overall, but:

- fetching additional code with git is bad, use additional distfiles
- libraries here are not meant for sharing, thus should not be put into /usr/lib but /usr/lib/root or similar, and (unless required for bindings, see below) should not be put into common/shlibs
- python and R bindings would preferably be subpackages rather than build options
- importing python modules from system interpreter doesn't work, they should be put into $py3_sitelib
- some fonts are nonfree shouldn't be packaged
- /usr/etc is a strange location

libRInside.so in common/shlibs looks fine

balejk commented:
Regarding the first point, it might be preferable to create individual packages for the respective components where it makes sense.

BenJarg commented:
@Chocimier Thanks for the feedback! I made the some changes:

-  I made the two packages which the make process downloads into destinct packages: xrootd and libvdt. All the other packages come with the ROOT source (if I followed the documentation correctly).
- It seems the libraries are needed by the R and Python bindings. But it also installs a bunch of other stuff with the .so files not directly needed by the bindings I don't think, so perhaps there's a way I can put some of the stuff into a designated directory, but it seems to all use the same CMake lib install directory.
- I made R and Python bindings subpackages. But now I get some issue with sonames not being found by the subpackages:
```
=> python3-pyroot-6.24.06_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libCore.so <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libgcc_s.so.1 <-> libgcc>=4.4.0_1
   SONAME: libpthread.so.0 <-> glibc>=2.32_1
   SONAME: libc.so.6 <-> glibc>=2.32_1
   SONAME: libTree.so <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libcppyy3_10.so <-> python3-pyroot (ignored)
   SONAME: libRIO.so <-> UNKNOWN PKG PLEASE FIX!
   SONAME: libcppyy_backend3_10.so <-> python3-pyroot (ignored)
=> ERROR: python3-pyroot-6.24.06_1: cannot guess required shlibs, aborting!
"
Where the libraries it can't find are provided by ROOT itself and installed into "/usr/lib". I'm pretty confused by this.
- The python bindings aren't put into the usual system python path, but that's set up when sourcing the "/usr/bin/thisroot.sh" which you need to do being using ROOT. I also added a "INSTALL.msg" saying that since it's not obvious. But it would make more sense for it to go there, so I send it to python site-packages now
- I missed the non-free fonts. Now I remove the Microsoft fonts and their license as a post_install. I have no idea if there is a more standard way to do that; let me know if there is.
- Yeah, it puts stuff in the "etc" directory of the install prefix, which is set to "/usr", so that's why it goes to "/usr/etc". Not sure if there is a way to override that in CMAKE? Or just move it manually?

balejk commented:
> Yeah, it puts stuff in the "etc" directory of the install prefix, which is set to "/usr", so that's why it goes to "/usr/etc". Not sure if there is a way to override that in CMAKE? Or just move it manually?

You can find the supported CMake variables [here](https://root.cern/install/build_from_source/#relevant-cmake-variables), I believe you are looking for `CMAKE_INSTALL_SYSCONFDIR`.

Based on [this](https://root.cern/install/build_from_source/#all-build-options) (expand `build options for ROOT 6.24`) I would say that there are more components which are downloaded automatically if no system version is found and some of them (such as `davix`) are enabled by default, however if I understand it correctly, git is not used for for example `davix` - this component in particular is fetched by CMake and checked against a checksum, so perhaps that's acceptable? I haven't checked the other components' download method.

Have you tried to build this on musl? Does the new version of xrootd build there? If not, you should probably restrict it to non-musl architectures in the template. Also, you probably have a wrong `homepage` set for the `xrootd` package.

BenJarg commented:
@balejk Thanks for pointing that out; I was only looking in the "builtin" options so I didn't see there are other things it may download. Also, right, it now installs stuff to "/etc/root/" instead.

Lastly, I forgot about the xrootd on musl problem. I just now tried to build it on a musl machine, and looks like it still doesn't build, so I restricted xrootd to glibc.

BenJarg commented:
libvdt on ARM architectures seemed to have some dopey issue where it wouldn't build with ninja, but it builds no problem with make (which I found after some trial and error). So I changed that. The ROOT subpackages are still complaining about some issue of not finding the sonames from the main ROOT package, so if the reason why seems obvious to anyone, let me know.

BenJarg commented:
Seems I had to add certain base libraries required by the subpackages to shlibs. Everything builds and installs properly now.

BenJarg commented:
Also, I had to restrict the `root-R` subpackage to glibc only because `R-cran-Rcpp` has a header file which the R part uses to compile that contains `#  include <execinfo.h>` which is only available on glibc.

Chocimier commented:
clad is still fetched during build, that needs to be fixed. git should be removed from build depends entirely.

BenJarg commented:
@Chocimier Right, to remove this problem, I turned off clad and now I could remove git with it still compiling.

Chocimier commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,94 @@
+build_options="fortran"
+build_options_default=$build_options
+python_version=3
+shlib_provides="libCore.so libTree.so libRIO.so libMatrix.so libThread.so libTMVA.so libMathCore.so"
```
common/shlibs is correct way to announce shared libs, no need for `shlib_provides=` in templates.

BenJarg replied:
Hm, when I didn't have that, it built right with xbps-src, but complained about missing, broken libraries when trying to install the subpackages with xbps-install. Maybe because those are in `/usr/lib/ROOT` and not in `/usr/lib`? I could perhaps move just those core libraries required by the subpackages into `/usr/lib`.

BenJarg replied:
Ok, symlinking the libraries needed to `/usr/lib` seems to solve this.

Chocimier replied:
You right, libraries go into package metadata only if exactly in `/usr/lib` or versioned.
Please keep shlib_provides= then.

Chocimier commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,94 @@
+
+python3-pyroot_package() {
+	short_desc="Python interface to ROOT"
+	depends="root>=${version}_${revision}"
```
Pleas add `python3` here.

Chocimier commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,94 @@
+distfiles="https://root.cern/download/root_v${version}.source.tar.gz"
+checksum=907f69f4baca1e4f30eeb4979598ca7599b6aa803ca046e80e25b6bbaa0ef522
+build_options="fortran"
+build_options_default=$build_options
```
Expand to `="fortran"`.

Chocimier commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,94 @@
+# Only i686 and x86_64 seem to be officially supported
+archs="i686* x86_64*"
+build_style=cmake
+configure_args="-Dgnuinstall=ON -Druntime_cxxmodules=OFF -Dclad=OFF $(vopt_if fortran '-Dfortran=ON') -DXROOTD_XrdCl_LIBRARY=/usr/lib/libXrdCl.so.3 -DCMAKE_INSTALL_LIBDIR=/usr/lib/root -DCMAKE_INSTALL_SYSCONFDIR='../etc/root' -DCMAKE_INSTALL_PYTHONDIR=${py3_sitelib/'usr/'}"
```
Wrap lines at approx. 80 columns.

BenJarg commented on srcpkgs/xrootd/template:
```diff
@@ -0,0 +1,13 @@
+pkgname=xrootd
+version=5.4.0
+revision=1
+archs="~*-musl"
```
It seems it's more complicated than that, and make take a bigger effort. After patching that, I also had to change a few `__GNUC__` s to `__GLIBC__`, and am still running into random compilation errors like `error: invalid conversion from 'int*' to 'socklen_t*'`. `xrootd` might be a little too tied to glibc.

BenJarg replied:
As for the `root-R` issue, it's because of one of `R-cran-Rcpp`'s header files the R bindings need to compile uses `execinfo.h`, which musl doesn't have.

Chocimier replied:
R-cran-Rcpp is fixed by #35301

BenJarg replied:
Great, the `root-R` part compiles on musl now without issue, so I removed the restriction.

BenJarg replied:
Ok, so I stumbled through throwing together a patch for xrootd to work on musl and pushed that here. I compiled and briefly tested the patched version on glibc and musl (on amd64). Someone let me know if anything looks clearly off. I could also send it upstream.

Chocimier commented on srcpkgs/xrootd/patches/musl.patch:
```diff
@@ -0,0 +1,303 @@
+         {sbP = statsQ.getBuff();
+          retc = recvfrom(udpFD, sbP->Data, sizeof(sbP->Data), 0,
+-                               &sbP->From.Addr, &fromLen);
++                               &sbP->From.Addr, (socklen_t*)&fromLen);
```
Defining `SOCKLEN_t` in `XrdSysPlatform.hh` more carefully would be better than casting `int*` to `socklen_t*`, both for upstream patch and for Void patch.

Chocimier reviewed (commented):
First, thanks for effort of patching it for musl.

I encourage you to send it upstream, even if maintainers will have a story for at least some of their `define`s.

BenJarg commented:
I opened a PR upstream, see xrootd/xrootd#1632.

BenJarg commented:
Just updated ROOT to the recent release after some testing. The new release needs now only a small patch to work on musl, which I've had pulled upstream: root-project/root#10045.

Chocimier commented:
Just to keep you updated: it is likely fine by now, just need to test it again before merge.

BenJarg commented:
The armv6l-musl build for xrootd was failing because of missing libatomic, so I added that for platforms without it. It's building correctly for armv6l-musl on my machine now.

Chocimier commented on srcpkgs/R-cran-RInside/template:
```diff
@@ -0,0 +1,12 @@
+# Template file for 'R-cran-RInside'
+pkgname=R-cran-RInside
+version=0.2.16
```
Please update, 0.2.16 source got removed already.

Chocimier commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,96 @@
+
+post_install() {
+	rm ${DESTDIR}/usr/share/root/fonts/LICENSE
+	rm ${DESTDIR}/usr/share/root/fonts/arialbd.ttf
```
`rm -v ${DESTDIR}/usr/share/root/fonts/[a-z]*.ttf` maybe.

BenJarg replied:
That's much cleaner. I made it `rm -v ${DESTDIR}/usr/share/root/fonts/[a-mt-z]*.ttf` to avoid removing one GPL font (symbol.ttf).

Chocimier reviewed (commented):
Some of your patches for musl were merged upstream but not present in release yet, can you add them here?

Works fine on glibc.

BenJarg commented on srcpkgs/root/template:
```diff
@@ -0,0 +1,96 @@
+ -Dclad=OFF $(vopt_if fortran '-Dfortran=ON')
+ -DXROOTD_XrdCl_LIBRARY=/usr/lib/libXrdCl.so.3
+ -DCMAKE_INSTALL_LIBDIR=/usr/lib/root -Dr=ON
+ -DCMAKE_INSTALL_SYSCONFDIR='../etc/root'
```
What would be the best place for these? `/usr/share`?

Chocimier replied:
`/usr/lib` (`libexec` ?), given that it contains precompiled headers and executable scripts.

BenJarg commented:
All those things should be fixed now, and I compiled and tested it on musl without issue.

q66 commented:
this package breaks qtcreator, because it results in it `shlib-depending` on `libCore.so`, which is now listed in `common/shlibs` and points to this package, even though it's irrelevant and only available on some platforms

@Chocimier

BenJarg commented:
@q66 Ok, what would be the easiest way to resolve this? I believe those libraries are there for the `root` sub packages.

Chocimier commented:
Oops sorry, haven't noticed as i have both installed.

Easiest way is to fix is to add `libCore.so` entry for qtcreator, will push later on.

q66 commented:
that seems pretty fragile though, because these names seem pretty generic/random and it's hard to tell when an accidental dependency is added
